[fusion] Add pattern constraint for dynamic slice fusion

shraiysh opened the pull request:
This patch constraints the dynamic-slice-fusion to only be triggered when the offset is either a constant or loop iteration offset that can guarentee that no D2H copy is required.

ezhulenev commented on xla/service/gpu/dynamic_slice_fusion_rewriter.cc:
```diff
@@ -137,6 +140,96 @@ bool IsAlignedSlice(const Shape& src_shape, const Shape& dst_shape,
+// Pattern matches the following IR (generated by `jax.lax.scan`) to check if
+// the offset is a loop iteration number:
+//
+//   param = (s32[], s32[], s32[16]{0}, s32[16]{0}) parameter(0)
```
can you please re-format the doc strict, put every instruction on a new line

ezhulenev commented on xla/service/gpu/dynamic_slice_fusion_rewriter.cc:
```diff
@@ -137,6 +140,96 @@ bool IsAlignedSlice(const Shape& src_shape, const Shape& dst_shape,
+      case PrimitiveType::S64:
+      case PrimitiveType::U32:
+      case PrimitiveType::U64:
+        VLOG(0) << "offset: " << offset.name() << " is a valid constant";
```
VLOG(0) - delete or bump to VLOG(2/3)

ezhulenev commented on xla/service/gpu/dynamic_slice_fusion_rewriter.cc:
```diff
@@ -137,6 +140,96 @@ bool IsAlignedSlice(const Shape& src_shape, const Shape& dst_shape,
+// constant or loop iteration offsets.
+bool HasConstantOrLoopIterationOffsets(
+    const HloDynamicIndexInstruction& instr) {
+  VLOG(0) << "Checking for instr : " << instr.name();
```
VLOG(0)

ezhulenev reviewed (commented):
there are more VLOG(0) lefovers

dimitar-asenov commented:
@ezhulenev Could you please have another look at the PR? Thanks.

shraiysh commented:
@ezhulenev can you please take a look at this pr again?

golechwierowicz commented:
@ezhulenev can you take a look?

ezhulenev commented:
There are couple of errors from linter:

```
CheckLint found errors.
Lines should be <= 80 columns wide  [whitespace/line_length] [2]
	xla/service/gpu/dynamic_slice_fusion_rewriter.cc:160](xla/service/gpu/dynamic_slice_fusion_rewriter.cc?l=160)
Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
	xla/service/gpu/dynamic_slice_fusion_rewriter.cc:161](xla/service/gpu/dynamic_slice_fusion_rewriter.cc?l=161)
Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
	xla/service/gpu/dynamic_slice_fusion_rewriter.cc:162](xla/service/gpu/dynamic_slice_fusion_rewriter.cc?l=162)
```

shraiysh commented:
> There are couple of errors from linter:

Actually, that was in the comment. I added `clang-format off` and `clang-format on` surrounding the code snippet, but for some reason it still flagged it. As a workaround, I moved the line. This should pass the linter now.

akuegel commented on xla/service/gpu/dynamic_slice_fusion_rewriter.cc:
```diff
@@ -44,6 +44,7 @@ limitations under the License.
 #include "xla/service/gpu/gpu_constants.h"
 #include "xla/service/gpu/hlo_traversal.h"
 #include "xla/service/gpu/ir_emission_utils.h"
+#include "xla/service/pattern_matcher.h"
```
You will also need to add a corresponding dependency to the BUILD file.

shraiysh replied:
Added it.

dimitar-asenov commented:
This PR breaks the multi-gpu variant of this test:
https://github.com/openxla/xla/blob/main/xla/service/gpu/fusions/BUILD#L177

Here is the error:

```
xla/service/gpu/fusions/dynamic_slice_fusion_test.cc:2499
Value of: changed
  Actual: false
Expected: true
```

Could you please have a look?

dimitar-asenov commented:
> This PR breaks the multi-gpu variant of this test: https://github.com/openxla/xla/blob/main/xla/service/gpu/fusions/BUILD#L177
> 
> Here is the error:
> 
> ```
> xla/service/gpu/fusions/dynamic_slice_fusion_test.cc:2499
> Value of: changed
>   Actual: false
> Expected: true
> ```
> 
> Could you please have a look?

The Bazel target that fails is `dynamic_slice_fusion_test_2gpu`.

shraiysh commented:
>This PR breaks the multi-gpu variant of this test:
https://github.com/openxla/xla/blob/main/xla/service/gpu/fusions/BUILD#L177

That test is expected to fail with this PR. This test relies on the fact that `DynamicSliceFusionRewriter` handles parameter. I will update this test.

golechwierowicz commented:
Hey, it still fails at `DynamicCustomCallSimple` test.

shraiysh commented:
Requesting review again. I have handled the failing testcase.
